analyzers/src/SonarAnalyzer.CFG/Roslyn/CfgAllPathValidator.cs

```diff
@@ -35,18 +35,34 @@ public abstract class CfgAllPathValidator
             this.cfg = cfg;
 
         public bool CheckAllPaths() =>
-            IsBlockOrAllSuccessorsValid(this.cfg.EntryBlock);
+            IsBlockOrAllSuccessorsValid(cfg.EntryBlock);
 
-        private bool IsBlockOrAllSuccessorsValid(BasicBlock block) =>
-            !IsInvalid(block) && (IsValid(block) || AreAllSuccessorsValid(block));
+        private bool IsBlockOrAllSuccessorsValid(BasicBlock block)
+        {
+            var isValid = !IsInvalid(block) && (IsValid(block) || AreAllSuccessorsValid(block));
+            visitedStatus[block] = !isValid;
+            return isValid;
+        }
 
         private bool AreAllSuccessorsValid(BasicBlock block)
         {
-            this.visited.Add(block);
-
-            return !block.SuccessorBlocks.Contains(cfg.ExitBlock)
-                   && block.SuccessorBlocks.Except(visited).Any()
-                   && block.SuccessorBlocks.Except(visited).All(IsBlockOrAllSuccessorsValid);
+            visitedStatus[block] = false; // protects from loops, don't fail the computation if hits itself
+            if (block.SuccessorBlocks.Any() && !block.SuccessorBlocks.Contains(cfg.ExitBlock))
+            {
+                foreach (var successorBlock in block.SuccessorBlocks)
+                {
+                    if ((visitedStatus.ContainsKey(successorBlock) && visitedStatus[successorBlock])
+                        || (!visitedStatus.ContainsKey(successorBlock) && !IsBlockOrAllSuccessorsValid(successorBlock)))
+                    {
+                        return false;
+                    }
+                }
+                return true;
```

Comment:
Check this out :) 
1) The `if` condition can be simplified as `visitedStatus.ContainsKey(successorBlock) ? visitedStatus[successorBlock] : !IsBlockOrAllSuccessorsValid(successorBlock)`
2) The `foreach` can be replaced with 
```
return !block.SuccessorBlocks.Any(x => visitedStatus.ContainsKey(x) ? visitedStatus[x] : !IsBlockOrAllSuccessorsValid(x));
```
3) We can merge the `if`s
4) We can swap `!Any` to `All` WHEN we align the state with the `isValid` like this `visitedStatus[block] = isValid;` (removing negation) 
```
return block.SuccessorBlocks.Any()
    && !block.SuccessorBlocks.Contains(cfg.ExitBlock)
    && block.SuccessorBlocks.All(x => x != cfg.ExitBlock && (visitedStatus.ContainsKey(x) ? visitedStatus[x] : IsBlockOrAllSuccessorsValid(x)));
```
5) _Optional:_ We can merge `Contains` and `All` actually
```
visitedStatus[block] = true; // protects from loops, don't fail the computation if hits itself
return block.SuccessorBlocks.Any()
    && block.SuccessorBlocks.All(x => x != cfg.ExitBlock && (visitedStatus.ContainsKey(x) ? visitedStatus[x] : IsBlockOrAllSuccessorsValid(x)));
```